integrations/nielsen-dtvr/lib/index.js

```diff
@@ -0,0 +1,382 @@
+'use strict';
+
+/**
+ * Module dependencies.
+ */
+
+var integration = require('@segment/analytics.js-integration');
+
+/**
+ * Expose `NielsenDTVR` integration.
+ */
+
+var NielsenDTVR = (module.exports = integration('Nielsen DTVR')
+  .option('appId', '')
+  .option('instanceName', '')
+  .option('id3Property', 'ID3')
+  .option('sendId3Events', [])
+  .option('optout', false)
+  .option('debug', false) // per Nielsen, customers should NOT ever enable in a production environment
+  .tag(
+    'http',
+    '<script src="http://cdn-gl.imrworldwide.com/conf/{{ appId }}.js#name={{ instanceName }}&ns=NOLBUNDLE">'
+  )
+  .tag(
+    'https',
+    '<script src="http:s//cdn-gl.imrworldwide.com/conf/{{ appId }}.js#name={{ instanceName }}&ns=NOLBUNDLE">'
+  ));
+
+/**
+ * Initialize.
+ *
+ * @api public
+ */
+
+NielsenDTVR.prototype.initialize = function() {
+  var protocol =
+    window.location.protocol === 'https:' ||
+    window.location.protocol === 'chrome-extension:'
+      ? 'https'
+      : 'http';
+  var config = {};
+  this.ID3 = null;
+  this.previousEvent = null;
+  this.isAd = false;
+
+  // Modified Nielsen snippet. It shouldn't load the Nielsen tag, but it should
+  // still successfully queue events fired before the tag loads.
+
+  /* eslint-disable */
+  !(function(t, n) {
+    t[n] = t[n] || {
+      // t = window, n = 'NOLBUNDLE'
+      nlsQ: function(e, o, c, r, s, i) {
+        // e = appId, o = instanceName, c = config
+        return (
+          (t[n][o] = t[n][o] || {
+            g: c || {},
+            ggPM: function(e, c, r, s, i) {
+              (t[n][o].q = t[n][o].q || []).push([e, c, r, s, i]);
+            }
+          }),
+          t[n][o]
+        );
+      }
+    };
+  })(window, 'NOLBUNDLE');
+  /* eslint-enable */
+
+  if (this.options.debug) config.nol_sdkDebug = 'debug';
+  this.client = window.NOLBUNDLE.nlsQ(
+    this.options.appId,
+    this.options.instanceName,
+    config
+  );
+
+  this.load(protocol, this.ready);
+};
+
+/**
+ * Loaded?
+ *
+ * @api public
+ * @return {boolean}
+ */
+
+NielsenDTVR.prototype.loaded = function() {
+  return this.client && typeof this.client.ggPM === 'function';
+};
+
+/**
+ * Track
+ *
+ * @api public
+ */
+
+NielsenDTVR.prototype.track = function(track) {
+  this.sendID3(track);
+};
+
+/**
+ * Video Content Started
+ *
+ * @api public
+ */
+
+NielsenDTVR.prototype.videoContentStarted = function(track) {
+  var time;
+  var adAssetId;
+  var metadata;
+  // Proactively ensure that we call "end" whenever new content
+  // starts. Here, we'll catch it if a customer forgets to call a Segment
+  // "Completed" event, so we'll end the video for them. `end` is also
+  // appropriate during a video interruption,
+  // e.g. if a user is alternating b/w watching two videos on the same page.
+  if (this.previousEvent && track !== this.previousEvent) {
+    if (this.previousEvent.proxy('properties.livestream') === true) {
+      time = this.previousEvent.timestamp().getUTCDate();
```

Comment:
@brennan - the call to `getUTCDate` will fail if `.timestamp` does not return a `Date` object. Might make sense to typecheck that return value or wrap this in a helper function of some sort.